File: backend/src/main/java/com/github/davidpotentini/comum/autorizacao/PermissaoService.java
package com.github.davidpotentini.comum.autorizacao;

import com.github.davidpotentini.enums.ENivel;
import com.github.davidpotentini.enums.ERecurso;
import com.github.davidpotentini.model.papeis.PapelPermissaoModel;
import com.github.davidpotentini.repository.papeis.PapelPermissaoRepository;
import org.springframework.stereotype.Service;

/** Negar por padrão: sem papel, ou recurso sem linha na matriz {@code PAPEL_PERMISSOES}, ⇒ negado. */
@Service
public class PermissaoService {

    private final PapelPermissaoRepository papelPermissaoRepository;

    public PermissaoService(PapelPermissaoRepository papelPermissaoRepository) {
        this.papelPermissaoRepository = papelPermissaoRepository;
    }

    public boolean permite(Long papCod, ERecurso recurso, ENivel minimo) {
        if (papCod == null) {
            return false;
        }
        return papelPermissaoRepository.findByPapCodAndRecurso(papCod, recurso)
                .map(PapelPermissaoModel::getNivel)
                .map(nivel -> rank(nivel) >= rank(minimo))
                .orElse(false);
    }

    /** Um nível cobre todos os abaixo dele. */
    private int rank(ENivel nivel) {
        return switch (nivel) {
            case NENHUM -> 0;
            case LEITURA -> 1;
            case EDICAO -> 2;
            case TOTAL -> 3;
        };
    }
}

File: frontend/src/app/core/autorizacao/permissao.service.ts
import { Injectable, computed, inject } from '@angular/core';
import { AuthService } from '../services/auth/auth.service';
import { ENivel, ERecurso } from '../../enums/autorizacao';

/** Ordem dos níveis: um nível "cobre" todos os abaixo dele. */
const RANK: Record<ENivel, number> = {
  [ENivel.Nenhum]: 0,
  [ENivel.Leitura]: 1,
  [ENivel.Edicao]: 2,
  [ENivel.Total]: 3,
};

/**
 * Substitui o `roleGuard` antigo (baseado em papéis fixos ADMIN_*). Lê a matriz
 * `permissoes` do papel logado (papel × recurso) e responde se o usuário atinge o
 * nível mínimo exigido para um recurso.
 */
@Injectable({ providedIn: 'root' })
export class PermissaoService {
  private readonly auth = inject(AuthService);

  readonly permissoes = computed(() => this.auth.sessao()?.permissoes ?? {});

  permite(recurso: ERecurso, minimo: ENivel = ENivel.Leitura): boolean {
    const atual = this.permissoes()[recurso] ?? ENivel.Nenhum;
    return RANK[atual] >= RANK[minimo];
  }
}
